File: tests/sheet_bench.py
"""Fit the peptide table the cartoon rebuilds backbones with, and score the frames.

The cartoon needs to know which way a strand faces. C-alpha curvature does not
say: it is dominated by the pleat, and it knows nothing about the neighbouring
strand, so two hydrogen-bonded residues ended up with their ribbon faces 38.9 deg
apart (median, over the natives) and sheets visibly failed to stack edge to edge.
The frames built here bring that to 21.2 deg, which is the floor: the sheets
themselves twist 20.0 deg between paired residues.

What does say is the backbone - a strand's carbonyls point at the strand it
pairs with, and its hydrogen bonds say which strand that is. py2Dmol only holds
C-alphas, so the backbone is REBUILT from the trace, PULCHRA-style (Rotkiewicz &
Skolnick 2008): bin the local internal coordinates, look up the peptide. PULCHRA
looks up a whole fragment and superimposes it; only C and N are needed here (O
follows from sp2 geometry), so it is one table read per residue.

This script fits that table - PEPTIDE_TABLE in viewer-cartoon.js - and reports
what it is worth on held-out chains. tests/ss_bench.js scores the secondary
structure that DSSP then derives from the rebuilt backbone, and
tests/sheet_bench.js scores the strand frames built on the sheet.

    python tests/sheet_bench.py --build   # natives.zip -> tests/out/sheet_truth.json
    python tests/sheet_bench.py --fit     # print the PEPTIDE_TABLE block
    node tests/sheet_bench.js             # score the frames

natives.zip (151 chains) is the same set ss_bench.py uses and is not in the
repo; put it in the repo root. Chains are split by index: even ones fit the
table, odd ones score it, so nothing is ever scored on its own training data.
"""
import argparse
import io
import json
import os
import zipfile

import numpy as np

_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
_OUT = os.path.join(_ROOT, "tests", "out")
NATIVES = os.path.join(_ROOT, "natives.zip")
TRUTH = os.path.join(_OUT, "sheet_truth.json")

BACKBONE = ("N", "CA", "C", "O")
# H-bond and ladder geometry. O...N under 3.3 A is the usual generous cut; the
# C-alpha window then throws out the i,i+4 helical bonds that also pass it.
HB_MAX = 3.3
LADDER_SEQ_MIN = 4
LADDER_CA_MIN = 4.2
LADDER_CA_MAX = 6.0


def _unit(v):
    return v / np.maximum(np.linalg.norm(v, axis=-1, keepdims=True), 1e-9)


def parse_natives():
    """-> [{name, ca, c, o, n, ladders}], residues with a full backbone only."""
    if not os.path.exists(NATIVES):
        raise SystemExit(
            f"{NATIVES} not found. It is 3.8 MB of native PDBs (the ss_bench set); "
            "put it in the repo root.")
    out = []
    with zipfile.ZipFile(NATIVES) as z:
        for name in sorted(z.namelist()):
            if not name.endswith(".pdb"):
                continue
            residues, order = {}, []
            for line in io.TextIOWrapper(z.open(name)):
                if not line.startswith("ATOM"):
                    continue
                atom = line[12:16].strip()
                if atom not in BACKBONE:
                    continue
                key = (line[21], line[22:27])
                if key not in residues:
                    residues[key] = {}
                    order.append(key)
                residues[key][atom] = (
                    float(line[30:38]), float(line[38:46]), float(line[46:54]))
            by_chain = {}
            for key in order:
                res = residues[key]
                if all(a in res for a in BACKBONE):
                    by_chain.setdefault(key[0], []).append(res)
            for chain, res in by_chain.items():
                if len(res) < 40:
                    continue
                ca = np.array([r["CA"] for r in res])
                c = np.array([r["C"] for r in res])
                o = np.array([r["O"] for r in res])
                nn = np.array([r["N"] for r in res])
                out.append(dict(
                    name=f"{os.path.basename(name)[:-4]}_{chain}",
                    ca=ca.tolist(),
                    c=c.tolist(),
                    o=o.tolist(),
                    n=nn.tolist(),
                    ladders=_ladders(ca, o, nn)))
    return out


def _ladders(ca, o, n):
    """Backbone H-bonded pairs that look like a beta ladder rung, not a turn."""
    pairs = set()
    for i in range(len(ca)):
        d = np.linalg.norm(n - o[i], axis=1)
        for j in np.flatnonzero(d < HB_MAX):
            if abs(int(j) - i) < LADDER_SEQ_MIN:
                continue
            a, b = min(i, int(j)), max(i, int(j))
            dca = np.linalg.norm(ca[a] - ca[b])
            if LADDER_CA_MIN < dca < LADDER_CA_MAX:
                pairs.add((a, b))
    return sorted(pairs)


# --- the model -------------------------------------------------------------
#
# Local frame on the peptide bond i -> i+1: u along CA(i)->CA(i+1), w normal to
# the CA(i-1),CA(i),CA(i+1) plane, v = w x u. Keys are PULCHRA's internal
# coordinates: r13 = |CA(i+1) - CA(i-1)|, and r14 = |CA(i+2) - CA(i-1)| signed
# by the chirality of the four. r13 separates helix from strand, r14 carries the
# handedness, and between them the peptide orientation is nearly determined.
# PULCHRA also bins the NEXT r13; held out it is worth 0.6 deg here, so it is
# dropped and the table stays one plane. Stored per bin: the offsets of C(i) and
# N(i+1) from CA(i) in that frame. O is not stored - the carbonyl carbon is sp2,
# so O sits in the CA-C-N plane opposite the bisector, 1.231 A out.
NB13, NB14 = 8, 25
R13_LO, R13_STEP = 4.6, 0.4
R14_LO, R14_STEP = -11.0, 0.9
CO_LENGTH = 1.231
MIN_PER_BIN = 5
CA_BOND_MIN, CA_BOND_MAX = 3.0, 4.2


def local_frames(ca):
    """-> (indices, local basis rows, r13, signed r14) for usable residues."""
    idx, frames, r13s, r14s = [], [], [], []
    for i in range(1, len(ca) - 2):
        a = ca[i + 1] - ca[i]
        length = np.linalg.norm(a)
        if not CA_BOND_MIN < length < CA_BOND_MAX:
            continue
        u = a / length
        w = np.cross(u, ca[i - 1] - ca[i])
        wl = np.linalg.norm(w)
        if wl < 1e-6:
            continue
        w = w / wl
        idx.append(i)
        frames.append(np.stack([u, np.cross(w, u), w]))
        r13s.append(np.linalg.norm(ca[i + 1] - ca[i - 1]))
        sign = np.sign(np.dot(np.cross(ca[i] - ca[i - 1], ca[i + 1] - ca[i]),
                              ca[i + 2] - ca[i + 1])) or 1.0
        r14s.append(sign * np.linalg.norm(ca[i + 2] - ca[i - 1]))
    return (np.array(idx, int), np.array(frames).reshape(-1, 3, 3),
            np.array(r13s), np.array(r14s))


def bins_of(r13, r14):
    b13 = np.clip(((r13 - R13_LO) / R13_STEP).astype(int), 0, NB13 - 1)
    b14 = np.clip(((r14 - R14_LO) / R14_STEP).astype(int), 0, NB14 - 1)
    return b13, b14


def fit(chains):
    """-> (grid, cells populated, samples). grid is [NB13, NB14, 6]: C then N."""
    local, b13, b14 = [], [], []
    for ch in chains:
        ca = np.array(ch["ca"])
        c = np.array(ch["c"])
        nn = np.array(ch["n"])
        idx, frames, r13, r14 = local_frames(ca)
        if not len(idx):
            continue
        local.append(np.concatenate([
            np.einsum("nij,nj->ni", frames, c[idx] - ca[idx]),
            np.einsum("nij,nj->ni", frames, nn[idx + 1] - ca[idx])], axis=1))
        a, b = bins_of(r13, r14)
        b13.append(a)
        b14.append(b)
    local = np.concatenate(local)
    b13 = np.concatenate(b13)
    b14 = np.concatenate(b14)
    grid = np.zeros((NB13, NB14, 6))
    count = np.zeros((NB13, NB14), int)
    np.add.at(grid, (b13, b14), local)
    np.add.at(count, (b13, b14), 1)
    filled = count >= MIN_PER_BIN
    grid[filled] /= count[filled][:, None]
    # Empty cells take their nearest filled neighbour so the renderer's lookup
    # is one index with no fallback path. r13 mismatch is weighted double: it is
    # the coarser coordinate and confusing helix with strand is the worse error.
    where = np.argwhere(filled)
    for i in range(NB13):
        for j in range(NB14):
            if filled[i, j]:
                continue
            d = np.abs(where[:, 0] - i) * 2 + np.abs(where[:, 1] - j)
            grid[i, j] = grid[tuple(where[np.argmin(d)])]
    return grid, filled.sum(), len(local)


def rebuild(ca, grid):
    """-> predicted C, O, N arrays (NaN where the frame is undefined)."""
    idx, frames, r13, r14 = local_frames(ca)
    c = np.full_like(ca, np.nan)
    o = np.full_like(ca, np.nan)
    n = np.full_like(ca, np.nan)
    if not len(idx):
        return c, o, n
    b13, b14 = bins_of(r13, r14)
    rows = grid[b13, b14]
    c[idx] = ca[idx] + np.einsum("nji,nj->ni", frames, rows[:, 0:3])
    n[idx + 1] = ca[idx] + np.einsum("nji,nj->ni", frames, rows[:, 3:6])
    # O opposite the CA-C / N-C bisector, in their plane
    bis = _unit(_unit(ca[idx] - c[idx]) + _unit(n[idx + 1] - c[idx]))
    o[idx] = c[idx] - bis * CO_LENGTH
    return c, o, n


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--build", action="store_true",
                    help="parse natives.zip -> tests/out/sheet_truth.json")
    ap.add_argument("--fit", action="store_true",
                    help="fit the table and print the PEPTIDE_TABLE block")
    args = ap.parse_args()
    if not (args.build or args.fit):
        ap.print_help()
        return
    if args.build or not os.path.exists(TRUTH):
        chains = parse_natives()
        for k, ch in enumerate(chains):
            ch["split"] = "fit" if k % 2 == 0 else "score"
        os.makedirs(_OUT, exist_ok=True)
        with open(TRUTH, "w") as fh:
            json.dump(chains, fh)
        print(f"wrote {TRUTH}: {len(chains)} chains, "
              f"{sum(len(c['ca']) for c in chains)} residues, "
              f"{sum(len(c['ladders']) for c in chains)} ladder rungs")
    if args.fit:
        with open(TRUTH) as fh:
            chains = json.load(fh)
        train = [c for c in chains if c["split"] == "fit"]
        grid, nfilled, nsamples = fit(train)
        errC, errN, errCO = [], [], []
        for ch in (c for c in chains if c["split"] == "score"):
            ca = np.array(ch["ca"])
            c_pred, o_pred, n_pred = rebuild(ca, grid)
            ok = ~np.isnan(c_pred[:, 0])
            errC.append(np.linalg.norm(c_pred[ok] - np.array(ch["c"])[ok], axis=1))
            okn = ~np.isnan(n_pred[:, 0])
            errN.append(np.linalg.norm(n_pred[okn] - np.array(ch["n"])[okn], axis=1))
            oko = ~np.isnan(o_pred[:, 0])
            errCO.append(np.degrees(np.arccos(np.clip(np.sum(
                _unit(o_pred[oko] - c_pred[oko])
                * _unit(np.array(ch["o"])[oko] - np.array(ch["c"])[oko]), axis=1), -1, 1))))
        errC = np.concatenate(errC)
        errN = np.concatenate(errN)
        errCO = np.concatenate(errCO)
        print(f"fit on {len(train)} chains / {nsamples} peptides, "
              f"{nfilled} of {NB13 * NB14} cells populated")
        print(f"held out: C {np.sqrt((errC ** 2).mean()):.2f} A rms, "
              f"N {np.sqrt((errN ** 2).mean()):.2f} A rms, "
              f"C=O direction {np.median(errCO):.1f} deg median")
        print("  (node tests/ss_bench.js scores the DSSP that runs on this backbone,")
        print("   node tests/sheet_bench.js the strand frames built from it)")
        flat = np.round(grid.reshape(-1), 3)
        body = ", ".join(f"{x:g}" for x in flat)
        import textwrap
        print("\n// paste into viewer-cartoon.js as PEPTIDE_TABLE:\n")
        print("\n".join("        " + line for line in textwrap.wrap(body, 92)))


if __name__ == "__main__":
    main()

File: tests/sheet_bench.js
/* Score the strand frames the cartoon draws sheets with.
 *
 *   python tests/sheet_bench.py --build    # once, writes tests/out/sheet_truth.json
 *   node tests/sheet_bench.js              # report
 *   node tests/sheet_bench.js --sweep      # try other smoothing/relaxation settings
 *
 * Two numbers, both in degrees, both over the residues of real beta ladders
 * (backbone H-bonds from the true coordinates, so they do not depend on any
 * prediction):
 *
 *   partner face   angle between the ribbon faces of two paired residues. This
 *                  is "do neighbouring strands stack edge to edge" - the thing
 *                  that was visibly wrong when strand frames came from C-alpha
 *                  curvature.
 *   strand twist   angle between the faces of consecutive residues in a strand.
 *                  This is "does the ribbon roll along its own length".
 *
 * Loads viewer-cartoon.js directly and calls the SAME exported functions the
 * renderer calls, so this cannot score a stale reimplementation.
 */
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const TRUTH = path.resolve(__dirname, 'out/sheet_truth.json');
const SRC = path.resolve(__dirname, '../py2Dmol/resources/viewer-cartoon.js');

function loadCartoon() {
    const sandbox = {
        window: { addEventListener() {}, dispatchEvent() {} },
        document: { createElement: () => ({ getContext: () => null }) },
        console,
        performance: { now: () => Date.now() },
        Event: function Event() {},
    };
    sandbox.window.window = sandbox.window;
    vm.createContext(sandbox);
    vm.runInContext(fs.readFileSync(SRC, 'utf8'), sandbox, { filename: SRC });
    return sandbox.window.py2dmolCartoon;
}

const sub = (a, b) => [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
const cross = (a, b) => [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]];
const dot = (a, b) => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
const unit = (a) => { const m = Math.hypot(a[0], a[1], a[2]); return m > 1e-9 ? [a[0] / m, a[1] / m, a[2] / m] : null; };
const angDeg = (a, b) => (a && b) ? Math.acos(Math.min(1, Math.abs(dot(a, b)))) * 180 / Math.PI : null;
const med = (a) => (a.length ? a.slice().sort((x, y) => x - y)[a.length >> 1] : NaN);
const p90 = (a) => (a.length ? a.slice().sort((x, y) => x - y)[Math.floor(a.length * 0.9)] : NaN);

function tangents(ca) {
    const n = ca.length, t = [];
    for (let i = 0; i < n; i++) {
        t.push(unit(sub(ca[Math.min(n - 1, i + 1)], ca[Math.max(0, i - 1)])) || [1, 0, 0]);
    }
    return t;
}

/** The frame the renderer used before the sheet pass: tangent x curvature. */
function curvatureSides(ca, tan) {
    const n = ca.length, out = [];
    for (let i = 0; i < n; i++) {
        const a = ca[Math.max(0, i - 1)], b = ca[i], c = ca[Math.min(n - 1, i + 1)];
        out.push(unit(cross(tan[i], [a[0] + c[0] - 2 * b[0], a[1] + c[1] - 2 * b[1], a[2] + c[2] - 2 * b[2]])));
    }
    return out;
}

function faces(sides, tan) {
    return sides.map((s, i) => (s ? unit(cross(tan[i], s)) : null));
}

/* Tangents of a flattened strand centreline - PyMOL's cartoon_flat_sheets, and
 * what the renderer's own sheet_flat does. The drawn plate's face is the
 * tangent crossed with the side vector, so a pleated centreline tilts the face
 * whatever frame it is handed: measuring against the raw trace scores the pleat
 * rather than the frame. */
function flatTangents(ca, sec, cycles) {
    const n = ca.length;
    const pv = ca.map((p) => p.slice());
    for (let c = 0; c < cycles; c++) {
        const tmp = pv.map((p) => p.slice());
        for (let i = 1; i < n - 1; i++) {
            if (sec[i] !== 'E' || sec[i - 1] !== 'E' || sec[i + 1] !== 'E') continue;
            for (let k = 0; k < 3; k++) tmp[i][k] = (pv[i - 1][k] + pv[i][k] + pv[i + 1][k]) / 3;
        }
        for (let i = 0; i < n; i++) pv[i] = tmp[i];
    }
    return tangents(pv);
}

function measure(sides, tan, sec, ladders) {
    const f = faces(sides, tan);
    const pf = [], tw = [];
    for (const [i, j] of ladders) {
        const a = angDeg(f[i], f[j]);
        if (a != null) pf.push(a);
    }
    for (let i = 0; i + 1 < sec.length; i++) {
        if (sec[i] !== 'E' || sec[i + 1] !== 'E') continue;
        const a = angDeg(f[i], f[i + 1]);
        if (a != null) tw.push(a);
    }
    return { pf, tw };
}

function run(api, chains, opts) {
    const pf = [], tw = [];
    for (const ch of chains) {
        const ca = ch.ca;
        const n = ca.length;
        const coords = ca.map((p) => ({ x: p[0], y: p[1], z: p[2] }));
        const { sec } = api.assignSecondary(coords, n, null);
        // ladders come from the TRUE backbone, not from the prediction, so the
        // score is about the frames and not about the pairing
        const truth = ch.ladders.filter(([i, j]) => sec[i] === 'E' && sec[j] === 'E');
        if (truth.length < 5) continue;
        const tan = tangents(ca);
        let sides;
        if (opts.curvature) {
            sides = curvatureSides(ca, tan);
        } else {
            const local = api.buildSheetFrames(coords, n, sec, null, truth, opts).local;
            sides = [];
            for (let i = 0; i < n; i++) {
                const o = i * 3;
                if (!local[o] && !local[o + 1] && !local[o + 2]) { sides.push(null); continue; }
                const fr = [];
                if (!api.localFrame((k) => coords[k], n, i, fr)) { sides.push(null); continue; }
                sides.push(unit([
                    fr[0] * local[o] + fr[3] * local[o + 1] + fr[6] * local[o + 2],
                    fr[1] * local[o] + fr[4] * local[o + 1] + fr[7] * local[o + 2],
                    fr[2] * local[o] + fr[5] * local[o + 1] + fr[8] * local[o + 2]]));
            }
        }
        const m = measure(sides, opts.flatTangent ? flatTangents(ca, sec, 4) : tan, sec, truth);
        pf.push(...m.pf);
        tw.push(...m.tw);
    }
    return { pf, tw };
}

(function main() {
    if (!fs.existsSync(TRUTH)) {
        console.error('missing ' + TRUTH + ' - run: python tests/sheet_bench.py --build');
        process.exit(1);
    }
    const api = loadCartoon();
    if (!api || !api.buildSheetFrames) {
        console.error('viewer-cartoon.js did not export buildSheetFrames');
        process.exit(1);
    }
    const chains = JSON.parse(fs.readFileSync(TRUTH, 'utf8'))
        .filter((c) => c.split === 'score' && c.ladders.length >= 10);
    // Floor: how far apart the faces of two paired residues are when both are
    // taken from the TRUE local sheet plane (the least-squares plane through the
    // six C-alphas i-1..i+1, j-1..j+1). Nothing built per residue can beat this,
    // because it is the sheet's own twist between the two.
    {
        const pf = [];
        for (const ch of chains) {
            const ca = ch.ca, n = ca.length;
            const coords = ca.map((p) => ({ x: p[0], y: p[1], z: p[2] }));
            const { sec } = api.assignSecondary(coords, n, null);
            const tan = tangents(ca);
            for (const [i, j] of ch.ladders) {
                if (sec[i] !== 'E' || sec[j] !== 'E') continue;
                // local sheet normal at i and at j, each from its own strand
                // tangent crossed with the direction to the partner
                const di = unit(sub(ca[j], ca[i]));
                const dj = unit(sub(ca[i], ca[j]));
                const ni = di && unit(cross(tan[i], di));
                const nj = dj && unit(cross(tan[j], dj));
                const a = angDeg(ni, nj);
                if (a != null) pf.push(a);
            }
        }
        console.log(`(floor: the sheet's own twist between paired residues, `
            + `median ${med(pf).toFixed(1)} p90 ${p90(pf).toFixed(1)})\n`);
    }
    const variants = [
        { name: 'curvature (old)', opts: { curvature: true } },
        { name: 'sheet frame (shipped)', opts: {} },
        { name: '  ... on a flattened strand', opts: { flatTangent: true } },
        { name: 'curvature, flattened strand', opts: { curvature: true, flatTangent: true } },
    ];
    if (process.argv.includes('--sweep')) {
        for (const relax of [16, 24, 40]) {
            for (const w of [1, 2, 3]) {
                variants.push({
                    name: `sweeps ${relax} neighbour weight ${w}`,
                    opts: { relax, along: w, across: w },
                });
            }
        }
    }
    console.log(`${chains.length} chains with beta ladders\n`);
    console.log('frame                          partner face      strand twist');
    console.log('                               median   p90     median   p90');
    for (const v of variants) {
        const { pf, tw } = run(api, chains, v.opts);
        console.log(`${v.name.padEnd(30)} ${med(pf).toFixed(1).padStart(6)} ${p90(pf).toFixed(1).padStart(6)}`
            + `   ${med(tw).toFixed(1).padStart(6)} ${p90(tw).toFixed(1).padStart(6)}`);
    }
})();
